components/engine_tirocks/src/perf_context.rs

```diff
@@ -0,0 +1,670 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{fmt::Debug, marker::PhantomData, mem, ops::Sub, time::Duration};
+
+use derive_more::{Add, AddAssign, Sub, SubAssign};
+use lazy_static::lazy_static;
+use prometheus::*;
+use prometheus_static_metric::*;
+use slog_derive::KV;
+use tikv_util::time::Instant;
+use tirocks::perf_context::{set_perf_flags, set_perf_level, PerfContext, PerfFlag, PerfFlags};
+use tracker::{Tracker, TrackerToken, GLOBAL_TRACKERS};
+
+use crate::{util, RocksEngine};
+
+macro_rules! report_write_perf_context {
+    ($ctx:expr, $metric:ident) => {
+        if $ctx.perf_level != engine_traits::PerfLevel::Disable {
+            $ctx.write = WritePerfContext::capture();
+            observe_write_time!($ctx, $metric, write_wal_time);
+            observe_write_time!($ctx, $metric, write_memtable_time);
+            observe_write_time!($ctx, $metric, db_mutex_lock_nanos);
+            observe_write_time!($ctx, $metric, pre_and_post_process);
+            observe_write_time!($ctx, $metric, write_thread_wait);
+            observe_write_time!($ctx, $metric, write_scheduling_flushes_compactions_time);
+            observe_write_time!($ctx, $metric, db_condition_wait_nanos);
+            observe_write_time!($ctx, $metric, write_delay_time);
+        }
+    };
+}
+
+macro_rules! observe_write_time {
+    ($ctx:expr, $metric:expr, $v:ident) => {
+        $metric.$v.observe(($ctx.write.$v) as f64 / 1e9);
+    };
+}
+
+make_auto_flush_static_metric! {
+    pub label_enum PerfContextType {
+        write_wal_time,
+        write_delay_time,
+        write_scheduling_flushes_compactions_time,
+        db_condition_wait_nanos,
+        write_memtable_time,
+        pre_and_post_process,
+        write_thread_wait,
+        db_mutex_lock_nanos,
+    }
+
+    pub struct PerfContextTimeDuration : LocalHistogram {
+        "type" => PerfContextType
+    }
+}
+
+lazy_static! {
+    pub static ref APPLY_PERF_CONTEXT_TIME_HISTOGRAM: HistogramVec = register_histogram_vec!(
```

Comment:
Why define these metrics in `tirocks`?

Reply:
It's defined in `engine_rocks`. I guess it's a partial work to hide the details of perf context.